microsetta_private_api/admin/admin_impl.py

```
@@ -282,6 +311,52 @@ def create_kits(body, token_info):
     return jsonify(kits), 201
 
 
+def add_barcodes_to_kits(body, token_info):
+    validate_admin_access(token_info)
+
+    kit_ids = body['kit_ids']
+    barcodes = body['barcodes']
+    generate_barcodes = body['generate_barcodes']
+
+    errors = []
+
+    with Transaction() as t:
+        kit_repo = KitRepo(t)
+        admin_repo = AdminRepo(t)
+
+        # First, make sure all of the Kit IDs are valid
+        for kit_id in kit_ids:
+            diag = kit_repo.get_kit_all_samples(kit_id)
+            if diag is None:
+                errors.append(f'Kit ID {kit_id} not found')
+
+        if generate_barcodes is False:
+            # Next, check that the Barcodes are unique if we're not generating
+            # novel ones
+            for barcode in barcodes:
+                diag = admin_repo.retrieve_diagnostics_by_barcode(barcode)
+                if diag is not None:
+                    errors.append(f'Barcode {barcode} already exists')
+
+            # And verify that the number of barcodes and kit IDs are equal
+            if len(barcodes) != len(kit_ids):
```

This test should be performed before the checks against the database under the fail fast principle

---

Given that this is an admin-facing tool with infrequent usage, I've opted to return a comprehensive set of errors so they can all be remedied at once, rather than simply returning the first error the API encounters. As such, the sequence of checks isn't important.